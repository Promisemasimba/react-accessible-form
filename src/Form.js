//@flow
import * as React from "react";
import classNames from "classnames";

type Props = {
    as?: React.ElementType,
    children: React.Node,
    className?: string,
    layout?: Layout,
    disabled?: boolean,
};

export type Layout = "stacked" | "aligned";

export const Form = React.forwardRef<Props, HTMLFormElement>((props: Props, ref) => {
    const {
        as = "form",
        className,
        children,
        disabled = false,
        layout = "stacked",
        ...rest
    } = props;
    const formContextValue = React.useMemo(
        () => ({
            disabled,
            layout,
        }),
        [disabled, layout]
    );
    //$FlowFixMe
    return React.createElement(
        as,
        { className: classNames(className, `form--${layout}`), ref, ...rest },
        <FormContext.Provider value={formContextValue}>{children}</FormContext.Provider>
    );
});

Form.displayName = "Form";

export type FormContextValue = {|
    disabled: boolean,
    layout: Layout,
|};

export const FormContext = React.createContext<FormContextValue>({
    disabled: false,
    layout: "stacked",
});
