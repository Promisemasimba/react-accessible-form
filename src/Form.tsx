import * as React from "react";
import classNames from "classnames";

import { FormControl } from "./FormControl";
import { FormGroup } from "./FormGroup";
import { FormLabel } from "./FormLabel";

type Props = {
    as?: React.ElementType;
    layout?: Layout;
    disabled?: boolean;
} & React.HTMLAttributes<HTMLFormElement>;

export type Layout = "stacked" | "aligned";

const FormComponent = React.forwardRef<HTMLFormElement, Props>((props: Props, ref) => {
    const {
        as = "form",
        className,
        children,
        disabled = false,
        layout = "stacked",
        ...rest
    } = props;
    const formContextValue = React.useMemo(
        () => ({
            disabled,
            layout,
        }),
        [disabled, layout]
    );
    return React.createElement(
        as,
        { className: classNames(className, `form--${layout}`), ref, ...rest },
        <FormContext.Provider value={formContextValue}>{children}</FormContext.Provider>
    );
});

FormComponent.displayName = "Form";

const Form = Object.assign(
    {
        Group: FormGroup,
        Label: FormLabel,
        Control: FormControl,
    },
    FormComponent
);

export type FormContextValue = {
    disabled: boolean;
    layout: Layout;
};

export const FormContext = React.createContext<FormContextValue>({
    disabled: false,
    layout: "stacked",
});

export { Form };
