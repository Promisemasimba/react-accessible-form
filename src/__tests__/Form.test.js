import * as React from "react";
import "jest-dom/extend-expect";
import { render } from "@testing-library/react";

import Form from "../index.js";

const getTestComponent = () => (
    <Form layout="aligned" onSubmit={(e) => e.preventDefault()}>
        <Form.Group id="1">
            <Form.Label>First Name</Form.Label>
            <Form.Control />
        </Form.Group>
        <Form.Group id="2">
            <Form.Label>Last Name</Form.Label>
            <Form.Control />
        </Form.Group>
        <Form.Group id="3">
            <Form.Label>Age</Form.Label>
            <Form.Control type="number" min="0" />
        </Form.Group>
        <Form.Group id="4">
            <Form.Label>Bio</Form.Label>
            <Form.Control as="textarea" />
        </Form.Group>
        <Form.Group id="5">
            <Form.Label>Email</Form.Label>
            <Form.Control type="email" />
            <small>You may occasionally receive marketing emails from us.</small>
        </Form.Group>
        <Form.Group id="6">
            <Form.Control type="checkbox" />
            <Form.Label>Receive Newsletter</Form.Label>
        </Form.Group>
        <Form.Group id="7">
            <Form.Label>Newsletter Frequency</Form.Label>
            <Form.Control as="select">
                <option>Daily</option>
                <option>Monthly</option>
                <option>Quarterly</option>
            </Form.Control>
        </Form.Group>
        <Form.Group id="8">
            <span />
            <Form.Control as="button">Submit</Form.Control>
        </Form.Group>
    </Form>
);

describe("Form", () => {
    it("renders and matches snapshot", () => {
        expect(() => {
            const { container } = render(getTestComponent());
            expect(container).toMatchSnapshot();
        }).not.toThrow();
    });
    it("applies layout classname", () => {
        const LAYOUT = "ABC";
        const { container } = render(<Form layout={LAYOUT} />);

        expect(container.querySelector(`.form--${LAYOUT}`)).not.toBe(null);
    });
});
