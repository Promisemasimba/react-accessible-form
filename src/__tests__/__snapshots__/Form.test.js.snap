// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders and matches snapshot 1`] = `
<div>
  <form
    class="form--aligned"
  >
    <div
      class="form__group form__group--aligned"
    >
      <div
        class="form__group__section"
      >
        <label
          class="form__label"
          for="1"
        >
          First Name
        </label>
      </div>
      <div
        class="form__group__section"
      >
        <input
          class="form__control"
          id="1"
        />
      </div>
    </div>
    <div
      class="form__group form__group--aligned"
    >
      <div
        class="form__group__section"
      >
        <label
          class="form__label"
          for="2"
        >
          Last Name
        </label>
      </div>
      <div
        class="form__group__section"
      >
        <input
          class="form__control"
          id="2"
        />
      </div>
    </div>
    <div
      class="form__group form__group--aligned"
    >
      <div
        class="form__group__section"
      >
        <label
          class="form__label"
          for="3"
        >
          Age
        </label>
      </div>
      <div
        class="form__group__section"
      >
        <input
          class="form__control"
          id="3"
          min="0"
          type="number"
        />
      </div>
    </div>
    <div
      class="form__group form__group--aligned"
    >
      <div
        class="form__group__section"
      >
        <label
          class="form__label"
          for="4"
        >
          Bio
        </label>
      </div>
      <div
        class="form__group__section"
      >
        <textarea
          class="form__control"
          id="4"
        />
      </div>
    </div>
    <div
      class="form__group form__group--aligned"
    >
      <div
        class="form__group__section"
      >
        <label
          class="form__label"
          for="5"
        >
          Email
        </label>
      </div>
      <div
        class="form__group__section"
      >
        <input
          class="form__control"
          id="5"
          type="email"
        />
        <small>
          You may occasionally receive marketing emails from us.
        </small>
      </div>
    </div>
    <div
      class="form__group form__group--aligned"
    >
      <div
        class="form__group__section"
      >
        <input
          class="form__control"
          id="6"
          type="checkbox"
        />
      </div>
      <div
        class="form__group__section"
      >
        <label
          class="form__label"
          for="6"
        >
          Receive Newsletter
        </label>
      </div>
    </div>
    <div
      class="form__group form__group--aligned"
    >
      <div
        class="form__group__section"
      >
        <label
          class="form__label"
          for="7"
        >
          Newsletter Frequency
        </label>
      </div>
      <div
        class="form__group__section"
      >
        <select
          class="form__control"
          id="7"
        >
          <option>
            Daily
          </option>
          <option>
            Monthly
          </option>
          <option>
            Quarterly
          </option>
        </select>
      </div>
    </div>
    <div
      class="form__group form__group--aligned"
    >
      <div
        class="form__group__section"
      >
        <span />
      </div>
      <div
        class="form__group__section"
      >
        <button
          class="form__control"
          id="8"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
</div>
`;
